### PR TITLE
Added in a pmt_vector_wrapper class

### PR DESCRIPTION
**Pull request by jsallay**
The idea is to allow for it to be accessed like an std::vector.  It should also be able to be used in places where a collection could be used such as std::sort.

This is a partial implementation.  The end goal would be to implement every function from std::vector that makes sense.  Likewise, we would do the same for pmt_map.

The key things that I am trying to show are:
A developer could use a pmt_vector_wrapper exactly like an std::vector.  If they are familiar with vectors (which they really should be) then they don't need to learn anything new to work with pmt vectors.
The smart pointer is wrapped by a class with useful functions.  There isn't awkward syntax for operators on a dereferenced pointer.  (For example it is weird to do (*ptr)[] on a smart pointer, 

Also note, that the constuctors of pmt_vector_wrapper and pmt_vector don't quite match up.  I followed the format of std::vector.  We would probably want to align these.

### CODE REVIEW

**Comment by mormj** [comment]
Syntactically this is a much nicer and more modern way to interface with the PMT vectors.  Unless I'm missing it, you probably want a sptr() method to return the shared pointer so it can be passed around like the previous implementation was

**Comment by mormj** [comment]
I'd like to get feedback from other folks, but this is definitely the right direction and it would be great to get the interface for the other PMT containers cleaned up as well.  

Once you get to the point of rebasing, let's stick with signed commits like we do in the GR codebase (according to the DCO) - which you can do with 'git commit --amend -s'.  I need to clean up the licensing throughout the codebase, but you can add your copyright information to the top of the files you add or significantly modify (via a spdx header).  I'll try and get those in place soon.

Also, when rebasing, the git commits should have a first line no longer than 50 chars and the more detailed description wrap at 72 characters.  

Thanks for taking this on and I think you're really moving it in the right direction.

File: pmt/include/pmt/pmtf_vector.hpp [hunk 1]
```diff
@@ -82,6 +82,58 @@ class pmt_vector : public pmt_base
...
+    using size_type = size_t;
+    using reference = T&;
+    using const_reference = const T&;
+    // Constructors.  Try to match std vector constructors.
+    // Right now I'm not allowing for a custom allocator, because I would believe 
+    // that we would say that flatbuffer is our allocator.
+    pmt_vector_wrapper(size_type n): 
+        d_ptr(pmt_vector<T>::make(std::vector<T>(n))) {}
+    pmt_vector_wrapper(size_type n, const value_type& val):
+        d_ptr(pmt_vector<T>::make(std::vector<T>(n))) {}
```
Comment (willcode) [inline pmt/include/pmt/pmtf_vector.hpp:100]: std::vector<T>(n, val)

**Comment by willcode** [comment]
I like the general idea. Are there other PMT structures (like dictionaries) that could be done this way too?

**Comment by jsallay** [comment]
> I like the general idea. Are there other PMT structures (like dictionaries) that could be done this way too?

Yes.  The idea really is to separate the interface from the underlying implementation. The current pmt implementation supports the concepts of vectors, linked lists, and maps. 

I don't see linked lists in the current implementation, but we could do this for anything that maps to an stl container.

File: pmt/include/pmt/pmtf_vector.hpp (hunk 1 above)
Comment (jsallay) [reply pmt/include/pmt/pmtf_vector.hpp:100]: Good catch. I'll fix that tonight.

**Comment by jsallay** [comment]
> Syntactically this is a much nicer and more modern way to interface with the PMT vectors. Unless I'm missing it, you probably want a sptr() method to return the shared pointer so it can be passed around like the previous implementation was

I can add in an sptr() method.  I understand that internally there is a desire to handle the pointers directly.  I assume for polymorphism.  I would hope that outside of that a developer would never need to have access to the pointer directly.  (I know I am late to the game on this, but if we used std::variant instead of polymorphism I believe that we could achieve the same effect, without exposing the pointer or really anything about the underlying implementation.  Though I readily admit that I am not really aware of how you are using the pointers.)

**Comment by jsallay** [comment]
> Once you get to the point of rebasing, let's stick with signed commits like we do in the GR codebase (according to the DCO) - which you can do with 'git commit --amend -s'. I need to clean up the licensing throughout the codebase, but you can add your copyright information to the top of the files you add or significantly modify (via a spdx header). I'll try and get those in place soon.
> 
> Also, when rebasing, the git commits should have a first line no longer than 50 chars and the more detailed description wrap at 72 characters.
> 
> Thanks for taking this on and I think you're really moving it in the right direction.

Thanks for the feedback.  I've only used git on internal projects before and never contributed to anything public like this.  I'm sure I'm going to make a lot of mistakes.  Just keep the feedback coming and I'll try to follow it.

**Comment by jsallay** [comment]
I'm working on the map.  In order to support the syntax of my_map["key"] = 4.0, I need a factory that will take in an arbitrary type and return the right kind of pmt object.  Does this exist already?  I'm not seeing it.  If not, I can write one.

**Comment by willcode** [comment]
`pmt::mp(whatever)` works in many cases (pmt_sugar.h).

**Comment by mormj** [comment]
> I can add in an sptr() method. I understand that internally there is a desire to handle the pointers directly. I assume for polymorphism.

Yes this is the current mechanism for polymorphism - just pass around a sptr to the base pmt class and then figure out what it is and cast it appropriately when it gets to its destination

**Comment by jsallay** [comment]
> `pmt::mp(whatever)` works in many cases (pmt_sugar.h).

I'm not finding pmt_sugar.h or an mp function anywhere in the repo (with find and grep).  Should I be looking on a particular branch maybe?

**Comment by willcode** [comment]
`gnuradio-runtime/include/pmt/pmt_sugar.h` is there in all versions. It is installed into `<prefix>/include/pmt`. If you can't find it in the repo, there's something wrong with your copy of the repo.

**Comment by jsallay** [comment]
I see now.  I was looking for something that worked with the new flatbuffer based pmts.

**Comment by willcode** [comment]
Gotcha. This would be a good addition to PMTF. `pmtf::pmt_scalar()` looks like it would work for all supported scalar types. But that leaves out strings, vectors and maps.

**Comment by jsallay** [comment]
I'll take a crack at writing a version that will return a smart pointer to the correct type for vector, map, sptr and scalar inputs.

**Comment by jsallay** [comment]
I have been trying in vain to write a wrapper class for the pmt map that will act like an stl map.  The big problem is that operator[] needs to return a modifiable reference into the map.  With the way that we are doing to the flatbuffers, that isn't possible directly.  I could make it work, but the code would be really ugly, and hard to follow and debug.  Here are my thoughts on options for a way forward:
1) Leave the pmtf_map class as is.  We can make other parts work like the stl but you won't be able to do things like:
  my_map["asdf"] = 4
  for (auto& k,v : my_map) {}

2) Change the pmtf_map class to be more accommodating.  Looking at the current interface, it looks like whenever we add or
modify any value in the map, we rebuild the whole flatbuffer from scratch in the set_value function.  This looks expensive to me and I would imaging wouldn't scale well. 

I did a tutorial with flatbuffers, but I'm by no means an expert with it.  So this suggestion may be terrible, but if it is possible for us to delay building the flatbuffer until we have to serialize the data, it would be more efficient to add/modify entries and it would lend itself to the stl map syntax.  Do you guys have any thoughts on the matter?

**Comment by mormj** [comment]
> I did a tutorial with flatbuffers, but I'm by no means an expert with it. So this suggestion may be terrible, but if it is possible for us to delay building the flatbuffer until we have to serialize the data, it would be more efficient to add/modify entries and it would lend itself to the stl map syntax. Do you guys have any thoughts on the matter?

The intention of leaving things serialized was so that we didn't have to maintain an underlying representation of the data - the serialized flatbuffer is the underlying data store.  Maps do get complicated since flatbuffers maintain 2 tables I think to represent a map, so there is more shuffling than a scalar or uniform vector.  But I'm completely open to different implementations here - the current code in there was a "just get it to work" to get some benchmark numbers.  I'm all for tearing it up and doing something completely different if it will be more functional and just as performant.

**Comment by mormj** [comment]
It would be really nice to have a `x["key"]` type of syntax

**Comment by jsallay** [comment]
I just pushed some proof of concept code that allows for the nicer syntax. I'll put in some notes here:
1. The data is stored in a map<string, pmt_wrap> (pmt_wrap is a thin wrapper around a pmt_sptr)
2. The class is pmt_map2.  If we like it, then I assume we will rename it to pmt_map
3. I added a virtual serialize_setup function to the pmt_base class.  This allows us to wait until we serialize to pack everything into the flatbuffer builder.
4. You can see the class in action in qa_pmt.cpp(MapWrapper)

Let me know your thoughts or if there are use cases that I am breaking here.

**Comment by mormj** [comment]
Sorry it has taken me a while to get back to this.

The pmt_map2 class is very nice - enabling the bracketed syntax is infinitely nicer than what I had in place before.  

I'd like to see the pmt_map2 class become the primary map class, as well as more qa functionality around it.  I can update the benchmark code (unless you want to) to compare the dict referencing and serializing performance of the new maps

Let's get the whole top level interface as your wrapped constructs so the user only sees std:: looking pmt objects.

This is all looking very good!

File: pmt/test/qa_pmt.cpp [hunk 2]
```diff
@@ -99,4 +101,63 @@ TEST(Pmt, VectorWrites)
...
+    pmt_vector_wrapper<uint32_t> vec4(vec3);
+    EXPECT_EQ(vec3.ptr(), vec4.ptr());
+}
+
+TEST(Pmt, MapWrapper) {
+    pmt_map2<std::string> x;
+    x["abc"] = 4;
+    x["qwer"] = std::vector<int>{1,2,4};
+    for (auto& [key, value]: x) {
+        std::cout << key << std::endl;
```
Comment (mormj) [inline pmt/test/qa_pmt.cpp:156]: Would be nice here to have some `EXPECT_EQ` tests and check against the value that is pulled out of the maps

File: pmt/test/qa_pmt.cpp [hunk 3]
```diff
@@ -99,4 +101,63 @@ TEST(Pmt, VectorWrites)
...
+    }
+}
+
+TEST(Pmt, PmtWrap) {
+    pmt_wrap x(4);
```
Comment (mormj) [inline pmt/test/qa_pmt.cpp:161]: Would `pmt_wrap` be then called just `pmt`?

File: pmt/test/qa_pmt.cpp [hunk 4]
```diff
@@ -99,4 +101,63 @@ TEST(Pmt, VectorWrites)
 
         EXPECT_EQ(int_pmt_vec->value(), int_vec_val_modified);
     }
-}
\ No newline at end of file
+}
+
+TEST(Pmt, VectorWrapper) {
+    pmt_vector_wrapper<uint32_t> x(10);
```
Comment (mormj) [inline pmt/test/qa_pmt.cpp:107]: I'm guessing `pmt_vector_wrapper` would become `pmt_vector` and `pmt_vector` would become something like `pmt_vector_base` ?

File: pmt/include/pmt/pmtf.hpp [hunk 5]
```diff
@@ -23,6 +23,7 @@ class pmt_base : public std::enable_shared_from_this<pmt_base>
 
     bool serialize(std::streambuf& sb)
     {
+        serialize_setup();
```
Comment (mormj) [inline pmt/include/pmt/pmtf.hpp:26]: There is a `bm_pmt_serialize_uvec.cpp` to do some limited measurements of serialization performance.  It will be good to keep our eye on the impact this has - not clear to me whether it will have any.

File: pmt/test/qa_pmt.cpp (hunk 2 above)
Comment (jsallay) [reply pmt/test/qa_pmt.cpp:156]: I can add that in.

File: pmt/test/qa_pmt.cpp (hunk 3 above)
Comment (jsallay) [reply pmt/test/qa_pmt.cpp:161]: Probably, but I need to think about that some more.

File: pmt/test/qa_pmt.cpp (hunk 4 above)
Comment (jsallay) [reply pmt/test/qa_pmt.cpp:107]: Yes

**Comment by mormj** [comment]
Hi @jsallay - wanted to check in on this feature and see if it is something that might be usable in the main GR codebase prior to our Sept. release?  We might be able to rally some other people to help with the coding and testing of it to continue your design here.

**Comment by mormj** [comment]
Closing, as PMT has moved here: https://github.com/gnuradio/pmt/
